### PR TITLE
feat: page, pageset, and pagecursor

### PR DESCRIPTION
**Pull request by rphmeier**

### CODE REVIEW

**Comment by rphmeier** [comment]
* **#29** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/29?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: core/src/page.rs [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        let i = (page_path.load_be::<u8>() as usize) << 1;
+        let base = (1 << (len + 1)) - 2;
+
+        (base + i, base + i + 1)
+    }
+}
+
+fn parent_page_id(_page_id: PageId) -> PageId {
+    // TODO
+    unimplemented!()
```
Comment (rphmeier) [inline core/src/page.rs:282]: depends on #28

File: core/src/page.rs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+}
+
+fn child_page_id(_page_id: PageId, _child_idx: u8) -> PageId {
```
Comment (rphmeier) [inline core/src/page.rs:285]: the real version of this function probably needs to fail when out of bounds.

File: core/src/page.rs [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+fn node_index(page_path: &BitSlice<u8, Msb0>) -> usize {
+    let len = page_path.len();
+    if len == 0 {
+        0
+    } else if len > DEPTH {
+        node_index(&page_path[..DEPTH])
+    } else {
+        let i = page_path.load_be::<u8>() as usize;
+        let base = (1 << len) - 2;
+        base + i
```
Comment (gabriele-0201) [inline core/src/page.rs:258]: I would move the `-2` here to emphasize that it is the 'conversion' between the number rule in the binary tree and how they are indexed in memory. I had to spend some time before figuring it out

File: core/src/page.rs [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    }
+}
+
+// Transform a bit-path to the indices of the two child positions in a page.
+//
+// The expected length of the page path is between 0 and `DEPTH - 1`, inclusive. All bits beyond
+// `DEPTH - 1` are ignored.
+fn child_node_indices(page_path: &BitSlice<u8, Msb0>) -> (usize, usize) {
```
Comment (gabriele-0201) [inline core/src/page.rs:266]: Why is this function not just?

```rust
let i = node_index(page_path) * 2;
(i, i+1)
```

File: core/src/page.rs [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    let len = page_path.len();
+    if len == 0 {
+        (0, 1)
+    } else if len > DEPTH - 1 {
+        child_node_indices(&page_path[..DEPTH - 1])
+    } else {
+        let i = (page_path.load_be::<u8>() as usize) << 1;
```
Comment (gabriele-0201) [inline core/src/page.rs:273]: I can't understand this shift, could you explain it?

File: core/src/page.rs (hunk 4 above)
Comment (rphmeier) [reply core/src/page.rs:261]: that logic would place multiple nodes at index 0

File: core/src/page.rs (hunk 5 above)
Comment (rphmeier) [reply core/src/page.rs:268]: It's just `* 2`, will modify for readability

File: core/src/page.rs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+/// The root page ID is fully zeroed.
+pub const ROOT_PAGE_ID: PageId = [0u8; 32];
+
+/// A read-only, borrowed page.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Page<'a> {
+    /// The underlying slice of nodes in memory.
+    /// This is expected to have a length equal to a power of two.
+    ///
+    /// The first `len - 2` slots are reserved for nodes. The last slot is the page ID.
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply core/src/page.rs:44]: Yeah, it's a backend thing which is used to resolve collisions in an on-disk hash map.

We could separate it (and indeed my first draft did exactly as you describe). I am about even on whether any approach is better. On one hand we know that the page layout in the real backend will include the page ID. On the other, that information isn't strictly needed at this level. May change,

but I will not block this PR on that refactoring.

File: core/src/page.rs (hunk 3 above)
Comment (rphmeier) [reply core/src/page.rs:253]: It's different than my mental model, which is that there is an offset for each length and an offset within each length - `base + i`, where `base = 2^len - 2` and `i` is the bit-string interpreted as a number.

**Pull request merged**